runtime/databricks/automl_runtime/forecast/pmdarima/model.py

```diff
@@ -0,0 +1,323 @@
+            )
+        # Validate the time range
+        pred_start_ds = min(df["ds"])
+        if pred_start_ds < self._start_ds:
+            raise MlflowException(
+                message=(
+                    f"Input time column '{self._time_col}' includes time earlier than "
+                    "the history data that the model was trained on."
+                ),
+                error_code=INVALID_PARAMETER_VALUE,
+            )
+        preds_pds = []
+        # Out-of-sample prediction if needed
+        horizon = int((max(df["ds"]) - self._end_ds) / pd.Timedelta(1, unit=self._frequency))
+        if horizon > 0:
+            future_pd = self._forecast(horizon)
+            preds_pds.append(future_pd)
+        # In-sample prediction if needed
+        if pred_start_ds <= self._end_ds:
+            in_sample_pd = self._predict_in_sample(start_ds=pred_start_ds, end_ds=self._end_ds)
+            preds_pds.append(in_sample_pd)
+        # Map predictions back to given time column
+        preds_pd = pd.concat(preds_pds).set_index("ds")
+        df = df.set_index("ds").join(preds_pd, how="left").reset_index()
+        return df["yhat"]
+
+    def _predict_in_sample(self, start_ds: pd.Timestamp = None, end_ds: pd.Timestamp = None):
+        if start_ds and end_ds:
+            start_idx = int((start_ds - self._start_ds) / pd.Timedelta(1, unit=self._frequency))
+            end_idx = int((end_ds - self._start_ds) / pd.Timedelta(1, unit=self._frequency))
+        else:
+            start_ds = self._start_ds
+            end_ds = self._end_ds
+            start_idx, end_idx = None, None
+        preds_in_sample, conf_in_sample = self.model().predict_in_sample(
+            start=start_idx, end=end_idx, return_conf_int=True)
+        dates_in_sample = pd.date_range(start=start_ds, end=end_ds, freq=self._frequency)
+        in_sample_pd = pd.DataFrame({'ds': dates_in_sample, 'yhat': preds_in_sample})
+        in_sample_pd[["yhat_lower", "yhat_upper"]] = conf_in_sample
+        return in_sample_pd
+
+    def _forecast(self, horizon: int = None):
+        horizon = horizon or self._horizon
+        preds, conf = self.model().predict(horizon, return_conf_int=True)
+        dates = pd.date_range(start=self._end_ds, periods=horizon + 1, freq=self._frequency)[1:]
+        preds_pd = pd.DataFrame({'ds': dates, 'yhat': preds})
+        preds_pd[["yhat_lower", "yhat_upper"]] = conf
+        return preds_pd
+
+    @staticmethod
+    def _validate_cols(df: pd.DataFrame, required_cols: List[str]):
+        df_cols = set(df.columns)
+        required_cols_set = set(required_cols)
+        if not required_cols_set.issubset(df_cols):
+            raise MlflowException(
+                message=(
+                    f"Input data columns '{list(df_cols)}' do not contain the required columns '{required_cols}'"
+                ),
+                error_code=INVALID_PARAMETER_VALUE,
+            )
+
+
+class MultiSeriesArimaModel(mlflow.pyfunc.PythonModel):
```

Comment:
Looks like there are quite some code duplication with uni-variate `ArimaModel`. Is there a way we can reduce that? e.g. create a list of `ArimaModel`s inside `MultiSeriesArimaModel` and use a for loop to call into the method inside.

Reply:
+1